src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Provide an overridable implementation of Dispose(bool) on {0} or mark the type as sealed. A call to Dispose(false) should only clean up native resources. A call to Dispose(true) should clean up both managed and native resources.
+            /// </summary>
+            private static void CheckProvideDisposeBoolRule(INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                context.ReportDiagnostic(type.CreateDiagnostic(ProvideDisposeBoolRule, type.Name));
+            }
+
+            /// <summary>
+            /// Checks rule: Ensure that {0} is declared as protected, virtual, and unsealed.
+            /// </summary>
+            private static void CheckDisposeBoolSignatureRule(IMethodSymbol method, INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                if (method.DeclaredAccessibility != Accessibility.Protected ||
+                    !(method.IsVirtual || method.IsAbstract || method.IsOverride) || method.IsSealed)
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeBoolSignatureRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(true), then calls GC.SuppressFinalize on the current object instance ('this' or 'Me' in Visual Basic), and then returns.
+            /// </summary>
+            private void CheckDisposeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                var validator = new DisposeImplementationValidator(suppressFinalizeMethod, type);
+                if (!validator.Validate(operationBlocks))
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeImplementationRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(false) and then returns.
+            /// </summary>
+            private static void CheckFinalizeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                // TODO: Implement check of Finalize
+            }
+
+            /// <summary>
+            /// Checks if type implements IDisposable interface or an interface inherited from IDisposable.
+            /// Only direct implementation is taken into account, implementation in base type is ignored.
+            /// </summary>
+            private bool ImplementsDisposableDirectly(ITypeSymbol type)
+            {
+                return type.Interfaces.Any(i => i.Inherits(disposableType));
```

Comment:
`type.AllInterfaces` returns also interfaces from inherited classes. This is not, what I want. I want to check, if class explicitly implements IDisposable, but not in base class.
For example class C inherits from class B and B implements IDisposable. In this case for type C `type.AllInterfaces` returns IDisposable. However, I don't care about class C, because it is obvious it does not violate any CA1063 rules, except maybe Dispose Override and Finalize Override.